### ISSUE TITLE
"msiDataObjRepl" replicating from a "normal" resource to a compound resource makes only one copy

### ISSUE THREAD

**Issue by cookie33** [irods/irods#2988]
Hi,

The following has been observed when replicating from a "normal" to a compound resource.

1. `iput` a file in a "normal" resource.
2. use `msiDataObjRepl` in a rule to replicate it to a compound resource with the parameter `all=`. Even if the compound resource has `auto_repl=on` the data is only replicated to the cache resource. NOT to the archive resource of the compound resource.

Using the icommand `irepl` to perform the same action causes the data to be replicated to cache and archive resource. So that's why I think the `msiDataObjRepl` should perform/behave the same. But it does not.

Greetings,

Robert Verkerk

**Comment by jasoncoposky** [comment]
i have tested this on both 4.1.7 tag and the master branch and both work as expected.  which version were you using?

**Comment by jasoncoposky** [comment]
my steps were:
```
~$ iadmin mkresc cache_resc unixfilesystem `hostname`:/tmp/irods/cache_resc
~$ iadmin mkresc arch_resc unixfilesystem `hostname`:/tmp/irods/arch_resc
~$ iadmin mkresc comp_resc compound
~$ iadmin addchildtoresc comp_resc cache_resc cache
~$ iadmin addchildtoresc comp_resc arch_resc archive
~$ iput -R demoResc VERSION.json file1
~$ irule -F /tmp/test_msiDataObjRepl.r
~$ ils -l
  rods              0 demoResc          455 2015-12-15.06:38 & file1
  rods              1 comp_resc;cache_resc          455 2015-12-15.06:38 & file1
  rods              2 comp_resc;arch_resc          455 2015-12-15.06:38 & file1
```

and my rule in /tmp:
```
~$ cat /tmp/test_msiDataObjRepl.r
test_msiDataObjRepl {
  *err = errormsg( msiDataObjRepl(*SourceFile,"destRescName=*Resource",*Status), *msg );
   if( 0 != *err ) {
       writeLine( "stdout", "*err - *msg" );
   }
}

INPUT *SourceFile="/tempZone/home/rods/file1", *Resource="comp_resc"
```

**Comment by cookie33** [comment]
Hi,

I was using 4.1.7. I will resets it with the example you provided and let the outcome be known.

Gr,

Robert

**Comment by cookie33** [comment]
Hi,
If you test it as described above it works. Sorry my description was not clear enough. It fails in an other way. It happens if the admin user tries to replicate an other users object:

```
iadmin mkresc cache_resc unixfilesystem `hostname`:/tmp/irods/cache_resc
iadmin mkresc arch_resc unixfilesystem `hostname`:/tmp/irods/arch_resc
iadmin mkresc comp_resc compound
iadmin addchildtoresc comp_resc cache_resc cache
iadmin addchildtoresc comp_resc arch_esc archive
mkdir -p /tmp/irods/cache_resc
mkdir -p /tmp/irods/arch_resc
chown -R <rodsuser>:<rodsuser>
```
As a non-priv user put a file:
```
robertv$ iput -R demoResc .bashrc  file1

robertv$ ils -l file1
  robertv           0 demoResc          124 2015-12-15.14:01 & file1
```
Now run the rule as the irods admin user:
```
rods$ irule -F /tmp/test_2.r
-818000 - execMicroService3: error when executing microservice
line 2, col 19
  *err = errormsg( msiDataObjRepl(*SourceFile,"destRescName=*Resource++++irodsAdmin=",*Status), *msg );
                   ^

caused by: DEBUG: msiDataObjRepl: rsDataObjRepl failed /alice/home/robertv/file1, status = -818000
```
and the file in iRODS:
```
rods$ ils -l /alice/home/robertv/file1
  robertv           0 demoResc          124 2015-12-15.14:01 & file1
  robertv           1 comp_resc;cache_resc          124 2015-12-15.14:20 & file1
```
And my test rule was:
```
rods$ cat /tmp/test_2.r
test_msiDataObjRepl {
  *err = errormsg( msiDataObjRepl(*SourceFile,"destRescName=*Resource++++irodsAdmin=",*Status), *msg );
   if( 0 != *err ) {
       writeLine( "stdout", "*err - *msg" );
   }
}

INPUT *SourceFile="/alice/home/robertv/file1", *Resource="comp_resc"
OUTPUT ruleExecOut
```
The error in the logfile is:
```
Dec 15 14:20:32 pid:11328 NOTICE: _rsGenQuery: genQuery status = -818000
Dec 15 14:20:32 pid:11328 NOTICE: rsGenQuery: _rsGenQuery failed, status = -818000
Dec 15 14:20:32 pid:11328 NOTICE: getDataObjInfo: rsGenQuery error, status = -818000
Dec 15 14:20:32 pid:11328 NOTICE: _rsDataObjRepl: getDataObjInfo for [/alice/home/robertv/file1] failed
Dec 15 14:20:32 pid:11328 NOTICE: rsDataObjRepl - Failed to replicate data object.
Dec 15 14:20:32 pid:11328 ERROR: [-]    iRODS/server/api/src/rsRegReplica.cpp:125:_call_file_modified_for_replica :  status [CAT_NO_ACCESS_PERMISSION]  errno [] -- message [_call_file_modified_for_replica - Failed to signal resource that the data object "/alice/home/robertv/file1" was registered]
        [-]     iRODS/server/drivers/src/fileDriver.cpp:723:fileModified :  status [CAT_NO_ACCESS_PERMISSION]  errno [] -- message [fileModified - Failed to call modified interface.]
                [-]     libcompound.cpp:496:repl_object :  status [CAT_NO_ACCESS_PERMISSION]  errno [] -- message [Failed to replicate the data object [/alice/home/robertv/file1] for operation [sync_object]]

Dec 15 14:20:32 pid:11328 NOTICE: _rsDataObjClose: RegReplica/ModDataObjMeta /alice/home/robertv/file1 err. stat = -818000
Dec 15 14:20:32 pid:11328 NOTICE: rsDataObjRepl - Failed to replicate data object.
Dec 15 14:20:32 pid:11328 DEBUG: msiDataObjRepl: rsDataObjRepl failed /alice/home/robertv/file1, status = -818000
Dec 15 14:20:32 pid:11328 NOTICE: readAndProcClientMsg: received disconnect msg from client
Dec 15 14:20:32 pid:11328 NOTICE: Agent exiting with status = 0
```

**Comment by cookie33** [comment]
What is also very bad is following: It thinks the files are there as the ampersand states. But only one is there.
```
14:43 irodstest1.storage.sara.nl:/var/lib/irods/iRODS/clients/icommands/test/rules3.0
rods$ ils -L /alice/home/robertv/file1
  robertv           0 demoResc          124 2015-12-15.14:01 & file1
        generic    /opt/Vault/home/robertv/file1
  robertv           1 comp_resc;cache_resc          124 2015-12-15.14:20 & file1
        generic    /tmp/irods/cache_resc/home/robertv/file1
```

But a check reveals that only one file is physically present:
```
14:57 irodstest1.storage.sara.nl:/var/lib/irods/iRODS/clients/icommands/test/rules3.0 
rods$ ls -l /opt/Vault/home/robertv/file1
-rw------- 1 rods rods 124 Dec 15 14:01 /opt/Vault/home/robertv/file1

14:57 irodstest1.storage.sara.nl:/var/lib/irods/iRODS/clients/icommands/test/rules3.0 
rods$ ls -l /tmp/irods/cache_resc/home/robertv/file1
ls: cannot access /tmp/irods/cache_resc/home/robertv/file1: No such file or directory
````
The one in the cache resource of the compound resource is NOT there.

**Comment by jasoncoposky** [comment]
The ADMIN_KW is not properly propagated, we will have it fixed for 4.1.8.
